client/src/components/LevelLayout.tsx

```diff
@@ -75,6 +99,113 @@ const LevelLayout: React.FC<Props> = ({
         {showInput && <StepValidation />}
         {children}
       </Container>
+
+      <Modal
+        isOpen={game.hasFailed}
+        onClose={() => {
+          navigate("/");
+        }}
+      >
+        <ModalOverlay />
+        <ModalContent>
+          <ModalHeader>You messed up too many times</ModalHeader>
+          <ModalBody>
+            <Text>
+              You used all 3 of your attempts. Do you want to restart this
+              level, try a previous level, or quit the game?
+            </Text>
+            {customRestart && (
+              <div>
+                <Select
+                  marginTop={4}
+                  onChange={(e) => setAlgo(parseInt(e.target.value))}
+                  defaultValue={"game"}
+                >
+                  {OPTIONS.map(({ value, name }, index) => (
+                    <option value={index} key={value}>
+                      {name}
+                    </option>
+                  ))}
+                </Select>
+                <Select
+                  onChange={(e) => setLevel(parseInt(e.target.value))}
+                  defaultValue={"0"}
+                  marginTop={2}
+                >
+                  {generateLevels()}
+                </Select>
+              </div>
+            )}
+          </ModalBody>
+          <ModalFooter>
+            {customRestart ? (
+              <Flex width={"100%"}>
+                <Button
+                  onClick={() => {
+                    setCustomRestart(false);
+                  }}
+                  variant="ghost"
+                >
+                  Easy Restart
+                </Button>
+                <Spacer />
+                <Button
+                  onClick={() => {
+                    game.jumpToLevel(level);
+                  }}
+                  colorScheme="blue"
+                  mr={1}
+                  variant="outline"
+                >
+                  Go to level
+                </Button>
+                <Button
+                  colorScheme="red"
+                  mr={3}
+                  onClick={() => {
+                    game.restartLevel();
+                    navigate("/");
+                  }}
+                >
+                  Quit
+                </Button>
```

Comment:
It seems there is a lot of duplicated code. Consider using the `customRestart` boolean at a lower granularity to reduce duplicate code.

Reply:
Very good point. Will keep the same for now, less coupling ;) easier to modify